graphql/src/execution/execution.rs

```diff
@@ -305,32 +314,37 @@ pub fn execute_root_selection_set(
 
     // Check if this query should be cached.
     if let (MaybeCached::Cached(cached), Some(key), Some(block_ptr)) = (&result, key, block_ptr) {
-        let mut cache = QUERY_CACHE.write().unwrap();
-
-        // If there is already a cache by the block of this query, just add it there.
-        if let Some(cache_by_block) = cache.iter_mut().find(|c| c.block == block_ptr) {
-            cache_by_block.cache.insert(key, cached.cheap_clone());
-        } else if *QUERY_CACHE_BLOCKS > 0 {
-            // We're creating a new `CacheByBlock` if:
-            // - There are none yet, this is the first query being cached, or
-            // - `block_ptr` is of higher or equal number than the most recent block in the cache.
-            // Otherwise this is a historical query which will not be cached.
-            let should_insert = match cache.iter().next() {
-                None => true,
-                Some(highest) if highest.block.number <= block_ptr.number => true,
-                Some(_) => false,
-            };
+        // Share errors from the herd cache, but don't store them in generational cache.
+        // In particular, there is a problem where asking for a block pointer beyond the chain
+        // head can cause the legitimate cache to be thrown out.
+        if cached.is_ok() {
```

Comment:
Style nit: We could avoid the extra indentation by moving this as an `Ok(_)` pattern match in the `if let` above.